### ISSUE TITLE
macOS: undefined SSL symbols in curl.a on x64

### ISSUE THREAD

**Issue by artificiel** [openframeworks/openFrameworks#7656]
i am building a project on a Mac M1 that must run on Intel, so turning"Build Native Architecture Only" to "NO". 

it's been working fine until "recently". I'm not compiling that constantly so I can't pinpoint more precisely but with git OF and a fresh download_libs it fails to link with a bunch of undefined symbols all relating to lib curl. 

for instance trying to compile imageLoaderWebExample:

<img width="293" alt="image" src="https://github.com/openframeworks/openFrameworks/assets/2172370/0314e021-e5c5-4bcb-8d2b-4df9fbf75d5b">

etc...

reverting `libs/curl/lib/osx/curl.a` to an earlier version (I had one dated April 20 2023) fixes the problem.

maybe "something else" is going on here but I unfortunately can't debug more proactively on intel right now but I wanted to raise this immediately.

**Comment by artificiel** [comment]
PS the lib seems legit fat:
```
% file curl.a
curl.a: Mach-O universal binary with 2 architectures: [x86_64:current ar archive] [arm64:current ar archive]
curl.a (for architecture x86_64):	current ar archive
curl.a (for architecture arm64):	current ar archive
```

**Comment by danoli3** [comment]
Looks like it was no successfully compiled with openssl

**Comment by ofTheo** [comment]
Hmm. Does this happen with the official 0.12.0 release too? 

As in the packaged version here: https://github.com/openframeworks/openFrameworks/releases/download/0.12.0/of_v0.12.0_osx_release.zip

Can you cause the issue with any of the examples?

**Comment by artificiel** [comment]
Above is with an example: imageLoaderWeExample

**Comment by artificiel** [comment]
will test later today with release

**Comment by ofTheo** [comment]
So just tried with the imageLoaderWebExample in 0.12.0 release and it works fine for me. 
But the nightly build does not and produces the errors you mention. 

Definitely a regression then.

**Comment by ofTheo** [comment]
Okay I think its tied to this change ( by me ;) ) 
https://github.com/openframeworks/apothecary/pull/280/files

<img width="1524" alt="image" src="https://github.com/openframeworks/openFrameworks/assets/144000/083e4870-7bba-47bc-b30f-3d8622d75427">

I can fix this as setting the deployment target to 10.15 or even 11.0 doesn't seem to make the error go away.

**Comment by artificiel** [comment]
yes i confirm on i7: fresh 0.12 works fine, the machine had a few months old git which i pulled to HEAD without running the udpdate_libs and it still worked; after updating libs it broke. notwithstanding fixing the issue, adding that example to the tests would be good to cover the curl stuff.

(i changed the title as the M1/non-native architecture thing is an irrelevant parameter)

**Comment by ofTheo** [comment]
still struggling with this issue. 
Essentially the Github runner ( macOS Arm ) doesn't have the openssl symbols for x86_64 installed, so our previous approach of doing 
`--with-dawinssl ` doesn't work for x86_64 

I think it probably links fine on an intel machine ( unconfirmed ) but building a universal app from M1 won't. 

Things I have tried:
- linking to the openssl version we are building (somehow I can't get make not to compile against my system ssl ) 
- installing homebrew and openssl and lib for both arm64 and intel and building / linking against that - works, but a massive pain in the butt. 

@artificiel did you say you found a way to build curl without openssl?

**Comment by artificiel** [comment]
did not try to build from scratch; my "solution" was to drop an old curl.a in the libs. i just tried on native i7 (13.4) with today's download_libs and unfortunately get the same linking errors.

my collection:

<img width="514" alt="image" src="https://github.com/openframeworks/openFrameworks/assets/2172370/af80d517-d96a-4344-8891-b8d5f0848634">

is curl.a solely utilized in OF for HTTP(S) downloads in ofURLFileLoader, or are there deeper considerations, or perhaps side-dependencies with other libs? because there are leaner alternatives such as the header-only https://github.com/yhirose/cpp-httplib — maybe the long term cost of supporting curl should be examined?

**Comment by danoli3** [comment]
Got a wip latest version, let me upload lib to test

**Comment by danoli3** [comment]
Updated macOS Libraries:

Open SSL 1.1.1w 
Curl 8.3.0 
x86_64 / arm64

x86_64 is library currently copied into standard linking location 

https://libs.danoli3.com/openssl-curl-patch-macOS.zip

**Comment by artificiel** [comment]
OK! 

removing curl.a from the Xcode settings dropping the libcurl.ca in "Link Binary with Libraries" produces a arm64+x64 app (built on M1 and run on M1 and i7). so I guess the apothecary process will make it so it lands in the same "curl.a" place?

**Comment by ofTheo** [comment]
Thanks @danoli3 ! 

I lipo-d the libs together as they weren't universal and tested them out but there was still a missing symbol for an Intel+arm build. 

`Undefined symbol: _SCDynamicStoreCopyProxies`

Adding 

```
-framework Foundation -framework SystemConfiguration
```

to: libs/openFrameworksCompiled/project/osx/CoreOF.xcconfig
fixed it but the example errors when run with this message:

```
[ error ] ofImage: loadImage(): couldn't load image from ""https://openframeworks.cc/about/0.jpg""
-1 SSL peer certificate or SSH remote key was not OK for request about
-1 SSL peer certificate or SSH remote key was not OK for request about
```

**Comment by ofTheo** [comment]
ps: the frustration of trying to fix a fairly small library led me to this thought last night:
https://github.com/openframeworks/openFrameworks/discussions/7562#discussioncomment-7060627

**Comment by artificiel** [comment]
i have another "diagonal" yet related question (still trying to grasp the apothecary approach). macOS provides libcurl -- are there specific OF reasons to recompile/repackage the functionality instead of linking against the OS? or is it about being homogenous across platforms? (that ultimately ties in my question above the API/function surface that curl.a is required to cover in the OF context).

(just tested by removing curl.a and instead link against libcurl.tbd from the + popup in "Link Libraries..." and imageLoaderWebExample works as a fat app, both machines running 13.4).

**Comment by artificiel** [comment]
> but there was still a missing symbol for an Intel+arm build.

hm strange it is not what happened here, my steps:
 - (on M1 13.4, current git/download_libs) 
 - PG new project from imageLoaderWebExample
 - try it, it runs
 - edit settings for all architectures // it misses symbols
 - remove definition of LIB_CURL path
 - drop the top-level libcurl.a from @danoli3's package in "link library" area
 - compile, runs (natively, not rosetta), and copy on i7 13.4, and runs

and alternatively
- remove libcurl.a // stops working
- click the [+] to add a system frameworks, and choose curl.tbd
- runs on M1 and i7

**Comment by artificiel** [comment]
maybe my methodology is not perfect — and this system is far from pristine, and both machines are running 13.4; perhaps there is more to it as to where the symbols come from...

**Comment by danoli3** [comment]
Oh yep this is just a bug in Curl it's fixed on master
https://github.com/curl/curl/commit/8b7cbe9decc205b08ec8258eb184c89a33e3084b

**Comment by ofTheo** [comment]
I just tried linking to system curl and it works great. 

only changes needed for a vanilla install are: 
[change this line](https://github.com/openframeworks/openFrameworks/blob/master/libs/openFrameworksCompiled/project/osx/CoreOF.xcconfig#L24) to:
`HEADER_CURL = ""`

and [change this line](https://github.com/openframeworks/openFrameworks/blob/master/libs/openFrameworksCompiled/project/osx/CoreOF.xcconfig#L40) to:
`LIB_CURL = "-lcurl"`

If this solves it for you @artificiel too and maybe on an older system @dimitre, could you test?

then maybe its an easy win?

**Comment by artificiel** [comment]
works! 

downloaded fresh git + libs on M1 13.4, made the above  changes, PG's FileWeb example, set to all architectures, compiled, runs, copied to i7 13.4, runs.

I'm not sure about signing etc but above testing a compile on older systems it might be good to also test a 13.4-compiled app on older systems (I left 10.15 as the target version, so I guess an intel running 10.15 would be the maximum test), I attach the build here. I did not touch signing settings so it might need xattr.

[imageLoaderWebExampleDebug.app.zip](https://github.com/openframeworks/openFrameworks/files/12676424/imageLoaderWebExampleDebug.app.zip)

**Comment by danoli3** [comment]
Solution is found by using frameworks instead of binaries as per macOS development guidelines.

**Comment by danoli3** [comment]
Probably more in theme to switch to frameworks anyway.... being open... LOL

**Comment by sebleedelisle** [comment]
Ah my apologies for the confusion, but it does actually work on the 0.12.0 release, the issue is only on the nightlies / GitHub.

**Comment by ofTheo** [comment]
closed by #7699
